### ISSUE TITLE
Support "dynamic" and writable DOM

### ISSUE THREAD

**Issue by justinushermawan** [dotnet/runtime#29690]
Does `JsonSerializer.Parse(String, Type, JsonSerializerOptions)` support for dynamic `ExpandoObject ` return type?

Something like this:
`dynamic p = JsonSerializer.Parse(json, typeof(ExpandoObject));`

**Comment by mauroservienti** [comment]
I read through the thread of comments and most are focused on deserialization, I'm facing an issue where also serialization of dynamic objects apparently silently "fails". In an attempt to reproduce my scenario I came up with the following minimal repro:

```
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.Text.Json;

namespace SampleConsoleApp
{
    class Program
    {
        static void Main(string[] args)
        {
            dynamic d = new CustomDynamicObject();
            d.AProperty = 10;
            var s = JsonSerializer.Serialize(d);

            Console.WriteLine(s);
            Console.Read();
        }
    }

    class CustomDynamicObject : DynamicObject 
    {
        private readonly IDictionary<string, object> properties = new Dictionary<string, object>();

        public override bool TryGetMember(GetMemberBinder binder, out object result) => properties.TryGetValue(binder.Name, out result);

        public override bool TrySetMember(SetMemberBinder binder, object value)
        {
            properties[binder.Name] = value;
            return true;
        }

        public override IEnumerable<string> GetDynamicMemberNames()
        {
            foreach (var item in properties.Keys)
            {
                yield return item;
            }
        }
    }
}

```

When run `s` is `{}`, so serialization doesn't fail but yields an empty json object.

Is this the right issue? or should I raise/follow a different one?

**Comment by ahsonkhan** [comment]
> Is this the right issue? or should I raise/follow a different one?

This is the right issue. No need to open one for the two halves of the same feature. This issue is about adding support for expando object (which means for both sides, serialize and deserialize).

**Comment by Lanayx** [comment]
I've met this issue today - wanted to say that `JsonElement` would work fine if it didn't depend on disponsed `JsonDocument`. One way I can think of going around this issue is to implement a destructor for `JsonDocument` so it's disposal can be postponed to later time - once all `JsonElement` objects are collected.

**Comment by zsalkus** [comment]
I needed dynamic object, however, it's not implemented yet. My solution was to use a dictionary.
`JsonSerializer.Deserialize<Dictionary<string, string>>(response)`
And then lookup for the key I need :)

**Comment by k3davis** [comment]
Just for my own sanity - is the issue scoped for 5.0 specifically ExpandoObject support or the ability to deserialize into a dynamic object? The conversation here doesn't all seem to match the issue title, and I'm definitely needing the latter. In my case I'm deserializing a nested dynamic, specifically `List<Dictionary<string, dynamic>>`. Switched back to Newtonsoft for now :(

**Comment by rs38** [comment]
just wanted to share some  c# 8 mods of @tchivs helpful code :)
 
```
/// <summary>
/// Temp Dynamic Converter with c# 8
/// by:tchivs@live.cn
/// </summary>
public class DynamicJsonConverter : JsonConverter<dynamic>
{
	public override dynamic Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
		=> reader.TokenType switch
		{
			JsonTokenType.True => true,
			JsonTokenType.False => false,
			JsonTokenType.Number => reader.TryGetInt64(out long l) ? 1 : reader.GetDouble(),
			JsonTokenType.String => reader.TryGetDateTime(out DateTime datetime) ? datetime.ToString() : reader.GetString(),
			JsonTokenType.StartObject =>  ReadObject(JsonDocument.ParseValue(ref reader).RootElement),
			// Use JsonElement as fallback.
				_ =>JsonDocument.ParseValue(ref reader).RootElement.Clone()
		};
	
	private object ReadObject(JsonElement jsonElement)
	{
		IDictionary<string, object> expandoObject = new ExpandoObject();
		foreach (var obj in jsonElement.EnumerateObject())
		{
			var k = obj.Name;
			var value = ReadValue(obj.Value);
			expandoObject[k] = value;
		}
		return expandoObject;
	}
	private object? ReadValue(JsonElement jsonElement)
		=>
		 jsonElement.ValueKind switch
		{
			JsonValueKind.Object => ReadObject(jsonElement),
			JsonValueKind.Array => ReadList(jsonElement),
			JsonValueKind.String => jsonElement.GetString(),
			JsonValueKind.Number =>  jsonElement.TryGetInt64(out long l) ? 1 :0,
			JsonValueKind.True => true,
			JsonValueKind.False =>false,
			JsonValueKind.Undefined => null,
			JsonValueKind.Null => null,
				_ => throw new ArgumentOutOfRangeException()
		};
	
	private object? ReadList(JsonElement jsonElement)
	{
		var list = new List<object?>();
		jsonElement.EnumerateArray().ToList().ForEach(j => list.Add(ReadValue(j)));
		return list.Count == 0 ? null : list;
	}
	
	public override void Write(Utf8JsonWriter writer,
		object value,
		JsonSerializerOptions options)
		{
		// writer.WriteStringValue(value.ToString());
		}
}
```

**Comment by ryan-hollister-q2** [comment]
@layomia thanks for the code here, was exactly what I needed. Wanted to point on a very subtle but important change needed. The two lines that parse the "Number" type are incorrect in your compressed version:

`JsonTokenType.Number => reader.TryGetInt64(out long l) ? 1 : reader.GetDouble(),`

should be 

`JsonTokenType.Number => reader.TryGetInt64(out long l) ? l : reader.GetDouble(),`

`JsonValueKind.Number =>  jsonElement.TryGetInt64(out long l) ? 1 :0,`

should be 

`JsonValueKind.Number =>  jsonElement.TryGetInt64(out long l) ? l :0,`

**Comment by xavierlaffargue** [comment]
@layomia  This is not serious. Support should have been done a long time ago (in fact System.Text.Json shouldn't have been launched without it in my opinion)! And we don't even have a deadline!

**Comment by SocVi100** [comment]
My scenario is about CosmosDB. I'm querying using the new .NET SDK wich uses this JsonSerializer and since it is a schemaless database I don't want to create a class for each projection of the database data I need to do (there are a bunch of different queries).
I need the results of the queries as lists of dynamic objects.

**Comment by RobbyDeLaet** [comment]
@SocVi100 Do not put your hopes on Microsoft for this one. Better stick with Newtonsoft's Json.net. @layomia has flushed all hopes for an near future integration. It seems they don't really care about the developer's needs. First they shouted "forget about json.net, we got you covered!" Luckily, developers didn't forget about json.net, or the alikes.

**Comment by rs38** [comment]
why don't you contribute? implement it and put in a PR!
https://github.com/dotnet/runtime/tree/master/src/libraries/System.Text.Json

**Comment by RobbyDeLaet** [comment]
> why don't you contribute? implement it and put in a PR!

I which I could.

**Comment by tjrobinson** [comment]
> My scenario is about CosmosDB. I'm querying using the new .NET SDK wich uses this JsonSerializer and since it is a schemaless database I don't want to create a class for each projection of the database data I need to do (there are a bunch of different queries).
> I need the results of the queries as lists of dynamic objects.

It's possible to configure a different serializer, using [CosmosClientBuilder.WithCustomSerializer](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.fluent.cosmosclientbuilder.withcustomserializer?view=azure-dotnet).

Here is an example: [CosmosJsonNetSerializer](https://gist.github.com/richstokoe/c82fc831c6b4020926f5b5f772f7cd70)

**Comment by ericstj** [comment]
We couldn't fit this feature into 5.0.  We had to prioritize along with the rest of the 5.0 feature work, and this one didn't fit.  FWIW this was very close to the cut line, thus the late move. 

> @layomia has flushed all hopes for an near future integration. It seems they don't really care about the developer's needs. First they shouted "forget about json.net, we got you covered!" Luckily, developers didn't forget about json.net, or the alikes.

@RobbyDeLaet can we try to keep this discussion constructive?  We are trying to do our best to respond to top requested customer features while maintaining design principles.  With respect to feature parity with Newtonsoft.Json, here's what we have to [say](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to).

> System.Text.Json focuses primarily on performance, security, and standards compliance. It has some key differences in default behavior and doesn't aim to have feature parity with Newtonsoft.Json. For some scenarios, System.Text.Json has no built-in functionality, but there are recommended workarounds. For other scenarios, workarounds are impractical. If your application depends on a missing feature, consider filing an issue to find out if support for your scenario can be added.

We don't aim to replace Newtonsoft.JSON.  If it works for you, continue to use it.  We'll do our best to make System.Text.Json as useful as possible while maintaining the gains we've achieved in performance, security, standardS compliance, and layering.  Over time we hope to make it work for as many folks as possible.   I am hearing the interest here and will make sure we focus on this moving forward.

**Comment by SocVi100** [comment]
Thanks for your suggestions. I've finally gone by the middle, using Newtonsoft serializer only for deserialization, but I still have to test. I hope it doesn't last too much to get ExpandoObject support on the default serializer of CosmosDB so I could get rid of the old one.
@RobbyDeLaet, thanks for your comment! I don't have much expectatives about Microsoft implementing anyting, I've felt fustrated too much times about them. As a freelance developer I've waited for years, literally, for essential features to be implemented on CosmosDB, EntityFramework and Identity infrastructure, and they make their own way ignoring us. It doesn't matter how many votes they have on the UserVoice or whatever. I supose that such features aren't needed for it's own developments so are out of scope...
The problem is always the same: Too much marketing = too much expectatives

**Comment by steveharter** [comment]
To help those that need this functionality, see the code samples above to help unblock:
- From @tchivs (which uses `ExpandoObject` for objects and `List<object>` for collections)
- From @denmitchell (which uses IL Emit for objects instead of `ExpandoObject`)

To help with requirements for this feature I will provide a new custom converter sample and will link that here when it's ready. After that I'll get that sample added to the [Newtonsoft work-around section](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to#table-of-differences-between-newtonsoftjson-and-systemtextjson). cc @tdykstra 

One detail not yet discussed is that supporting `dynamic` requires a reference to the very large `System.Linq.Expressions.dll` assembly; this complication may mean in the future we add the dynamic converter in a new out-of-band assembly (e.g. `System.Text.Json.Converters.dll`) to get pay-to-play for those that care about deployment size (such as a self-contained app or a Blazor client app).

**Comment by Julius-ERCM** [comment]
> Just for my own sanity - is the issue scoped for 5.0 specifically ExpandoObject support or the ability to deserialize into a dynamic object? The conversation here doesn't all seem to match the issue title, and I'm definitely needing the latter. In my case I'm deserializing a nested dynamic, specifically `List<Dictionary<string, dynamic>>`. Switched back to Newtonsoft for now :(

I am now testing an old function I implemented to deserialize nested objects and seeing the same issue noted in this topic with "valueKind".

Was there a way to fix this using the new netonsoft.json packages? Because the above code worked for the root level object but not for the nested objects.

**Comment by RobbyDeLaet** [comment]
@ericstj Apologies, my reaction was perhaps a bit too negative and harsh, but like @SocVi100 I'm a freelance developer and sometimes frustrations get the upper hand. But at least we got a discussion started here. Thank you for your response, now we got at least a clear message about the status of this request.

**Comment by steveharter** [comment]
> Just for my own sanity - is the issue scoped for 5.0 specifically ExpandoObject support or the ability to deserialize into a dynamic object

I assume the desired semantics are to have a dynamic type so after deserializing you can access all properties (including nested) in a late-bound manner:
```cs
dynamic obj = JsonSerializer.Deserialize<dynamic>(json);
string s = obj.MyChildProperty.MyStringList[2];
```

and I assume the desired semantics are also support `ExpandoObject` explicitly:
```cs
ExpandoObject expando = JsonSerializer.Deserialize<ExpandoObject>(json);
dynamic obj = expando;
string s = obj.MyChildProperty.MyStringList[2];
```

So my understanding of the scope:
- `Deserialize<dynamic>(json)` returns either a primitive, collection or object. An object here will likely be `ExpandoObject` but could be `IDynamicMetaObjectProvider` or a JITed type depending on implementation. This is different than today which always returns `JsonElement`.
- `Deserialize<ExpandoObject>(json)` (which implements `IDynamicMetaObjectProvider`) should create a proper `ExpandoObject` which in turn can be used with `dynamic`. This is different than today which only creates expando-properties for the root properties and `JsonElement` instances for all nested properties.
- `Serialize<ExpandoObect>()` works as expected
- `Serialize<IDynamicMetaObjectProvider`()> may or may not be implemented depending on whether there are scenarios where `ExpandoObject` is not used.

Semantics in 3.0 - 5.0:
- Serializing an `ExpandoObject` works since `ExpandoObject` implements `IDictionary<string, object>` and STJ will correctly serialize each `object` value whether it is a primitive, collection or object.
- Deserializing into an `ExpandoObject` sort of works but only the root properties are "proper" expando-properties; any nested properties will be `JsonElement` so there is inconsistency in the programming model thus deserializing `ExpandoObject` should be avoided unless a custom converter is used for that.

Options for 3.0-5.0:
- Write a custom converter for `ExpandoObject`, `dynamic` and\or `object` as mentioned in the posts above; I'll provide a link soon to a new sample that ideally maps to a 6.0 feature.
- If you only serialize, not deserialize, you can use `ExpandoObject` (or `dynamic` if the type is based on `ExpandoObject`). Since deserializing `ExpandoObject` has an inconsistency issue today in STJ, `ExpandoObject` is not recommended if deserializing.
- Use `JsonElement` instead of dynamic objects. `JsonElement` doesn't attempt to "guess" what type the JSON maps to, so you have to be explicit by calling GetString(), GetInt32(), etc.

Also, depending on implementation, the implementation of a custom converter likely means there is some "guessing" about what CLR types the JSON maps to during deserialization. For example, a JSON string may map to either a `DateTime` or a `string`, a JSON number may map to a `double` or a `long`, and the JSON array type needs to be determined. This needs to be vetted and compared against Newtonsoft semantics. This should also align to what type is returned when a property is of Type `object` (today it is `JsonElement`).

Here's some examples of 3.0 - 5.0 STJ semantics:

```cs
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.Diagnostics;
using System.Linq;
using System.Text.Json;

namespace ConsoleApp
{
    class Program
    {
        const string ExpectedJson = "{\"A\":\"A\",\"B\":[1,2],\"C\":42,\"D\":\"2020-01-01T00:00:00\",\"E\":{\"A_Child\":\"A_Child\"}}";

        static void Main(string[] args)
        {
            DateTime dateTime = new DateTime(2020, 1, 1);

            dynamic myDynamicChild = new ExpandoObject();
            myDynamicChild.A_Child = "A_Child";

            dynamic myDynamic = new ExpandoObject();
            myDynamic.A = "A";
            myDynamic.B = new List<int>() { 1, 2 };
            myDynamic.C = 42;
            myDynamic.D = dateTime;
            myDynamic.E = myDynamicChild;

            // Verify we can call late-bound property.
            int c = myDynamic.C;
            Debug.Assert(c == 42);

            // STJ can serialize with ExpandoObject since it implements IDictionary<string, object>.
            string json = JsonSerializer.Serialize<ExpandoObject>(myDynamic);
            Debug.Assert(json == ExpectedJson);

            // Using 'dynamic' against backing ExpandoObject works.
            json = JsonSerializer.Serialize<dynamic>(myDynamic);
            Debug.Assert(json == ExpectedJson);

            // Deserialize with <dynamic>, <object> and <JsonElement>.
            // For 5.0, using one of these is recommended over ExpandoObject because the programming model will be
            // consistent for the root type and all nested types.
            // Using <JsonElement> makes it clear and non-abiguous.
            // Using <object> by default uses 'JsonElement', but can be overridden by a custom converter.
            // Using <dynamic> uses 'object' which uses 'JsonElement'.
            {
                dynamic d = JsonSerializer.Deserialize<dynamic>(json);
                VerifyJsonElement(d);

                try
                {
                    // We will get an exception here if we try to access a dynamic property since 'object' is deserialized
                    // as a JsonElement and not an ExpandoObject.
                    c = d.C;
                    Debug.Fail("Should have thrown Exception!");
                }
                catch (Exception ex)
                {
                    Debug.Assert(ex.Message == "'System.Text.Json.JsonElement' does not contain a definition for 'C'");
                }

                // Serializing with <object> creates a JsonElement by default (can be changed by a custom converter).
                object o = JsonSerializer.Deserialize<object>(json);
                VerifyJsonElement((JsonElement)o);

                // Serialize with explicit <JsonElement>.
                JsonElement e = JsonSerializer.Deserialize<JsonElement>(json);
                VerifyJsonElement(e);
            }

            // Deserialize with ExpandoObject. This creates an ExpandoObject with the root Type having Expando-properties
            // but the value of those properties will be JsonElement. All other nested properties\objects\collections will
            // also be JsonElement. Due to the inconsistency of having only root-level Expando-properties (such as A_Child),
            // deserializing as ExpandoObject is not recommended (unless a ExpandoObject custom converter is used).
            {
                // When STJ deserializes, it creates ExpandoObjects via IDictionary<string, object> where 'object' is JsonElement.
                ExpandoObject expando = JsonSerializer.Deserialize<ExpandoObject>(json);
                Debug.Assert(((IDictionary<string, object>)expando).Keys.Count == 5);
                myDynamic = expando;

                JsonElement jsonElement = myDynamic.A;
                Debug.Assert(jsonElement.GetString() == "A");

                jsonElement = myDynamic.B;
                Debug.Assert(jsonElement.EnumerateArray().Count() == 2);

                jsonElement = myDynamic.C;
                Debug.Assert(jsonElement.GetInt32() == 42);

                jsonElement = myDynamic.D;
                Debug.Assert(jsonElement.GetDateTime() == dateTime);

                jsonElement = myDynamic.E;
                // Here we have an inconsistency. Nested object property must use JsonElement (not a dynamic property).
                Debug.Assert(jsonElement.GetProperty("A_Child").GetString() == "A_Child");

                // Re-serialize works as expected.
                json = JsonSerializer.Serialize<ExpandoObject>(myDynamic);
                Debug.Assert(json == ExpectedJson);

                // Re-serialize works as expected; dynamic works here since backed by ExpandoObject in this example.
                json = JsonSerializer.Serialize<dynamic>(myDynamic);
                Debug.Assert(json == ExpectedJson);
            }

            void VerifyJsonElement(JsonElement elem)
            {
                // Verify JsonElement
                Debug.Assert(elem.GetProperty("A").GetString() == "A");
                Debug.Assert(elem.GetProperty("B").EnumerateArray().Count() == 2);
                Debug.Assert(elem.GetProperty("C").GetInt32() == 42);
                Debug.Assert(elem.GetProperty("D").GetDateTime() == dateTime);
                Debug.Assert(elem.GetProperty("E").GetProperty("A_Child").GetString() == "A_Child");

                // Re-serialize
                json = JsonSerializer.Serialize<dynamic>(elem);
                Debug.Assert(json == ExpectedJson);

                json = JsonSerializer.Serialize<JsonElement>(elem);
                Debug.Assert(json == ExpectedJson);
            }
        }
    }
}
```

**Comment by TAGC** [comment]
@rs38 Can you fix your code snippet based on what @ryan-hollister-q2 pointed out?

**Comment by steveharter** [comment]
As promised, the PR providing a dynamic implementation sample is at https://github.com/dotnet/runtime/pull/42097.

**Comment by John0King** [comment]
I saw many people in this issue talk about the `dynamic` but what about the Writable JSON DOM ?   
As a Newtonsoft.Json user,  I won't use `dynamic`,  instead I  use `JToken`  , because in many case the json is `data: T | T[]` ,
and to handle those data, you can't use just `dynamic` because it hard to know what is the type of the object.  
the `JToken`  solve 99% of the serialize/deserialize problem ,  and we can "corrent" the `JToken` to right form (or add necessary other properties ), and then send to other place (eg.  return to client or write to db or some message handler)

**Comment by steveharter** [comment]
> I saw many people in this issue talk about the dynamic but what about the Writable JSON DOM ?

The proposal for 6.0 is to support for `dynamic` and non-`dynamic` cases with an overlapping but consistent programming model. Newtsoft does this as well since `JToken` implements `IDynamicMetaObjectProvider` which supports `dynamic`. This means if you don't want to use `dynamic`, you don't have to. In general, I would expect most consumers to not use `dynamic` since it is slower.

>  and we can "corrent" the JToken to right form

Can you elaborate on "correcting" to make sure we have the scenario. I assume you mean changing the DOM to modify a value of a property\element or by adding\removing properties\elements.

**Comment by John0King** [comment]
@steveharter 
> Can you elaborate on "correcting" to make sure we have the scenario. I assume you mean changing the DOM to modify a value of a property\element or by adding\removing properties\elements.

yes,  for example  change the `data:T|T[]`  to `data:T[]`  .  I can't do it (or hard) with `dynamic` because I need to know the type of the property/object ,  and `JToken` can do this perfectly because every property/object have a `TokenType` property.

but today the `JsonDocument`  `JsonElement`  is just a `StringReader` or `SpanReader`  (compare to newtonsoft.Json is `new JsonTextReader(new StringReader(ref jsonString))` , just that today we can't pass by `ref`) ,   
and  it seems that  this issue   became to use `dynamic`  to serialize and  deserialize the object again and again,  it will be way slower than `JToken` like  "C# object"  , and hard to detect it's type .

we don't need a `string reader`  we need the "C# object"   !

**Comment by keithn** [comment]
Seems weird C# doesn't have a standard library to manipulate JSON data, only serialize, deserialize from static types, or read only an arbitrary json document.  Really need to be able to read, query and to manipulate JsonDocument, add elements, remove elements, etc and then turn it back to a string.  The simple pain of getting any arbitrary piece of json and adding an extra element to it is astonishing.  Think you guys need to implement a library that allows full manipulation of JSON data, not just a means to and end for deserializing/serializing static types.

**Comment by rs38** [comment]
in .NET there are 2 standards: 3rd party Nuget package Json.NET and .NET Core’s built-in classes ```System.Text.JSON``` which are very fast but a bit limited compared to Json.NET.

here is an example of manipulating a JSON file without deserializing with the built in Json classes:
```
using var json = File.OpenRead (jsonPath);
using JsonDocument document = JsonDocument.Parse (json);
var options = new JsonWriterOptions { Indented = true };
using (var outputStream = File.Create ("NewFile.json"))
using (var writer = new Utf8JsonWriter (outputStream, options))
{
 writer.WriteStartArray();
 foreach (var person in document.RootElement.EnumerateArray())
 {
 int friendCount = person.GetProperty ("Friends").GetArrayLength();
 if (friendCount >= 2)
 person.WriteTo (writer);
 }
}
```
(from: C# 8.0 in a Nutshell)

**Comment by udlose** [comment]
it's unfortunate that the PR - https://github.com/dotnet/runtime/pull/42097 @steveharter added for an example won't load...

![image](https://user-images.githubusercontent.com/656118/112203306-65b18500-8be0-11eb-85ff-4d431f74e7ca.png)

**Comment by John0King** [comment]
Am I the only one who want a  JSON  POCO Object in  C# (eg.  `JToken`) ,  so  we can manipulate the "`object`" easily and also other Json library can directly handle it too.

and @rs38 you solution  is about write,  and a mutable JSON DOM  is about  change and read (or write)

```C#
//  do some change to the original JSON
var jobj = xxxx.Parse(Request.Body) as JObject;
if(ShouldInjectClaim == true)
{
  
   jobj.Add("Claims", JArray.Parse(new []{  "claim1". "claim2" }));
}

//  handle the "data"
DoSomething(jobj);  // mostly read again


```

**Comment by ericstj** [comment]
@John0King I don't think you're the only person, that's part of what this issue is about.  Have you seen https://github.com/dotnet/designs/pull/163?

@udlose not sure why GitHub is having trouble loading the PR.  You can always check out the commit: https://github.com/dotnet/runtime/commit/095539fd800775fc9542a07f411cbc1b7e72084e or even just browse to these tests that are now part of the product.

> I also noticed that the example is missing a ToString() implementation override on JsonDynamicString - was this intentional? 

I think that's just because the sample was minimal.  I see @steveharter mentions this in the design doc:
https://github.com/dotnet/designs/pull/163/files#diff-34fc37119c1d4491b844207aeed3a9c846c3e2d835fd712e164b1e47b89687c0R77

**Comment by steveharter** [comment]
Closing; the `JsonNode` feature is in main and will be in Preview 4.

Preview 5 will also add some new capabilities for `JsonObject` including deterministic property ordering and being able to use `JsonObject` as an "extension property" via `[JsonExtensionDataAttribute`]. These are mentioned in the [PR description](https://github.com/dotnet/runtime/pull/51025).

**Comment by udlose** [comment]
@steveharter , @ericstj  I pulled Steve's code sample for supporting `dynamic` from https://github.com/dotnet/runtime/commit/095539fd800775fc9542a07f411cbc1b7e72084e.  It worked fine when I was running under .NET Core 3.1.  I'm upgrading the app to .NET 5 and getting an error:

`JsonSerializerExtensions.JsonDynamicType.Value is inaccessible due to its protection level`

I can easily change these from `internal` to `public` but I'm concerned about why it broke when migrating to .NET 5.  Is there a new version of the code sample that works for .NET 5?

**Comment by ericstj** [comment]
It looks to me like a compiler error around a identifier resolution, where something in another assembly is resolving to that Value member when it didn't before.  It could be something unrelated to the update.  If you have an issue to report (even against sample code) it's better to open a new issue and provide us with full details and a repro, that way we can better reproduce and diagnose what you are seeing.

Also, if you're willing to grab a nightly build you can try the official feature which is in as of last week: https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet6&package=System.Text.Json&protocolType=NuGet&version=6.0.0-preview.5.21229.9

Thanks!

**Comment by techhowdy** [comment]
You can also try this
```
var obj = JsonDocument.Parse(Convert.ToString(your_dynamic_object));

 var propertyVal = obj.RootElement.GetProperty("Your_Property_Name").GetString();
```

**Comment by steveharter** [comment]
For those still interested in dynamic how valuable is that now that we have a writable DOM with `JsonNode`?

Please common on https://github.com/dotnet/runtime/issues/53195
